Workflow cache

BabyBoySnow opened the pull request:
After creating your pull request, tick these boxes if they are applicable to you.

- [x] I have tested my changes against the `fae` branch (the latest developmental version), and this pull request is targeting that branch as a base
- [x] I have tested my changes on Python 3.8 or higher
- [x] I have ensured my code is formatted using [Black](https://github.com/psf/black)

----

### Description
Adds improvements to run a faster workflow, and reduce overhead by caching the requirements. When a push or pull is done, and the workflow is triggered this should be able to grab a restore-key and use that. The cron job is set to run every Sunday at midnight, clearing the cache, and creating a new one to ensure packages are up to date. Each job and python version has a unique cache, by prefixing with: typing-Linux-pip-3.12 or lint-Linux-pip-3.10


### Related issues (if applicable)

BabyBoySnow committed:
cache pip packages

Adds improvements to run a faster workflow, and reduce overhead by caching the requirments. When a push or pull is done, and the workflow is triggered this should be able to grab a restore-key and use that. The cron job is set to run every sunday at midnight, clearing the cache, and creating a new one to ensure packages are up to date.

BabyBoySnow committed:
job specific caches

itsTheFae commented on .github/workflows/py-checks.yml:
```diff
@@ -1,15 +1,20 @@
+  push:
+  pull_request:
+  schedule:
+    - cron: '0 0 * * 0'  # Runs every Sunday at midnight to clear the cache
```
Could you elaborate on how scheduling this is clearing the cache?  
By my understanding the cache is cleared based on key invalidation, which only really seems to happen when the hash for the requirements file changes.

itsTheFae replied:
Actually, lets just remove the schedule all together.  Cache is cleared automatically by some server-side, LRU-like process.

itsTheFae commented on .github/workflows/py-checks.yml:
```diff
@@ -19,47 +24,64 @@ jobs:
+          key: lint-${{ runner.os }}-pip-${{ matrix.python-version }}-${{ hashFiles('requirements.dev.txt') }}-${{ github.run_id }}
+          restore-keys: |
+            lint-${{ runner.os }}-pip-${{ matrix.python-version }}-${{ hashFiles('requirements.dev.txt') }}-
+            lint-${{ runner.os }}-pip-${{ matrix.python-version }}-
```
I'm not sure we need the less strict key as a fall-back here.   Wouldn't this effectively make non-existing caches just go back and use the old existing cache?

itsTheFae commented on .github/workflows/py-checks.yml:
```diff
@@ -19,47 +24,64 @@ jobs:
+          key: typing-${{ runner.os }}-pip-${{ matrix.python-version }}-${{ hashFiles('requirements.dev.txt') }}-${{ github.run_id }}
+          restore-keys: |
+            typing-${{ runner.os }}-pip-${{ matrix.python-version }}-${{ hashFiles('requirements.dev.txt') }}-
+            typing-${{ runner.os }}-pip-${{ matrix.python-version }}-
```
Same here as on lines 38-39.  
Additionally, could we leverage the cache between these jobs?   (So as to avoid caching two copies of pip libs for py 3.10)

itsTheFae commented:
~~Just need to iron out some details around how cache works and if its data will bloat the repo.~~  
~~At a glance, it looks like we'll need another job / workflow to handle actually purging caches.  We should probably avoid doing that in triggered checks though.~~

Based on the job history on my branches, the recent changes without caching have saved around 20 seconds from typical job run time.

itsTheFae committed:
Use cache save/restore to run between jobs.

Update action versions to avoid node deprecation.
Cache is conditionally saved based on cache-miss.  The warning issued about save failure is not an issue.

itsTheFae commented:
The last commit here splits caching into save and restore steps, where saves depend on cache-miss.  
This might still generate a warning/annotation on the workflow when cache is created for the first time.  But since it is only a warning, and in almost every case they will be saving the same info, I think this is fine.
